### ISSUE TITLE
PR CI builds shouldn’t be vulnerable to “latest” distro instability 

### ISSUE THREAD

**Issue by simbo1905** [sobolevn/git-secret#235]
We have seen a number of CI builds fail while packaging on `*-latest` distros. This leads to a situation where the logic in a PR is good, passes on stable versions of each distro family we test on (Fedora, Debian, Alpine), but is broken on a “*-latest”.

This is a problem as it makes it harder to get working contributions onto master. Yet we can expect that the majority of our users are on stable versions of distros and that packages built on stable versions are forwards compatible and will install on the latest versions. 

We should have PR buillds only package on stable (long term support versions) of distros. That will prove that the contributed code is going to benefit the majority of our users immediately. 

We do need to do packaging on the latest of distros. We could have a build on master that runs everything. Then as soon as the latest distros make a breaking change we can see and fix it “little and often”. Else we can leave it until we tag for a release to do a full builds and do the release from the latest distros. That forces us to fix the problems when we release. 

The pros of “do latest packaging on master” are we get to see and fix problems as the distros break us. The cons are that our master will break often and that might be “bad publicity” and deter new contributions.

The pros of ”do latest packaging on release” is that we only see breaks upon releases. We will be motivated to fix them to get the release shipped. The cons are that we might hit a multiple issues on multiple distros due to “storing up trouble” between infrequent releases.

**Comment by joshrabinowitz** [comment]
I'm disabling tests on alpine-latest for now, so we can merge the #230 PR if we want, see 
https://github.com/sobolevn/git-secret/pull/231/commits/142944c575eec4358a61cf2260f590a799ddb39e

I don't care too much about this issue, as @simbo1905 points out there are pros and cons to both approaches. Dealing with broken builds at random moments during PR develop,ent does get old, though.

**Comment by joshrabinowitz** [comment]
@simbo1905 any more thoughts on this? I'm thinking we should move towards a 0.2.5 release not that we have fixes for spaces in filenames and paths, and a fix for 'running git-secret in a subdirectory'


[also @sobolevn ]

**Comment by simbo1905** [comment]
It looks like the PR to fix Alpine latest builds has worked [406921657](https://travis-ci.org/sobolevn/git-secret/jobs/406921657) [406921662](https://travis-ci.org/sobolevn/git-secret/jobs/406921662). Looking on hub.docker.com the [alpine latest tag](https://hub.docker.com/_/alpine/) is applied to Alpine 3.8. 

I suggest that for PR/master builds we test on specific versions of alpine, docker, ubuntu, and debian using both gpg1 and gpg2. Then have our release builds use the `:latest` tags of those distros. If we find that a release build fails, we should take a look at what version has the `:latest` tag, then bump the PR/master build to test that specific new version, fix it on master, then try the release again. 

The net effect will be that new features/fixes in PRs will be against tested stable versions so won't get broken by distro updates. When we do a release we will occasionally hit issues but will be motivated to address them to get the release out the door.

**Comment by simbo1905** [comment]
This is unblocked now that alpine latest works which is also alpine current. It should also look to fix #241 which I closed so that it won’t clash with this and so that we can make one set of changes to the ci bulld to cut a new release.

**Comment by simbo1905** [comment]
I suggest we retire the `rolling` Ubuntu build. That will be less stable than `latest`. With all distros (included centos) we should test against a stable `latest` when we release and against a named version for PRs and master.

**Comment by simbo1905** [comment]
Looking at the current [test suites](https://github.com/sobolevn/git-secret/blob/b504facc5d2fa7d2c7e8c8abdee86cff5835e6f8/.kitchen.yml#L120) it seems to do the `gnupg-git` suite on 3 out of 6 distros. It's not clear to me why it might exclude `ubuntu-latest` but not `ubuntu-rolling`. @sobolevn is there a particular design as to which platforms the `gnupg-git` suite runs?

**Comment by sobolevn** [comment]
It is better to ask @hurricanehrndz about it.

**Comment by joshrabinowitz** [comment]
@simbo1905 I assumed the logic was to use the `gnupg-git` suite on one the most recent version of each of distro tested on (debian/ubunto/fedora). 

I presumed
 * fedora was seen as representative of centos (maybe not valid),  and 
 * since we don't test on 'rolling' versions of debian, `latest` is used for that `gnupg-git` test.

I figured he was just trying to keep the overall build times down since the `gnupg-git` suite takes longer than  `gnupg1` and `gnupg2`.
